File: utilities/currency-exchange.js
const CONVERSION = {
    "copper": {
        "platinum": 1000,
        "gold": 100,
        "electrum": 50,
        "silver": 10,
    },
    "silver": {
        "platinum": 100,
        "gold": 10,
        "electrum": 5,
    },
    "electrum": {
        "platinum": 20,
        "gold": 2,
    },
    "gold": {
        "platinum": 10,
    },
}

const CURRENCY_NAMES = {
    "cp": "copper",
    "sp": "silver",
    "ep": "electrum",
    "gp": "gold",
    "pp": "platinum",
    "copper": "copper",
    "silver": "silver",
    "electrum": "electrum",
    "gold": "gold",
    "platinum": "platinum",
};

const ARG_PATTERN = /(\d+[a-z]+)/g

class CurrencyConverter {

    constructor(args) {
        const inventory = {};
        const currencyDetails = args.join('').match(ARG_PATTERN);
        currencyDetails.forEach((currencyDetail) => {
            const currencyLongName = CURRENCY_NAMES[Object.keys(CURRENCY_NAMES).find((currencyName) => currencyDetail.includes(currencyName))];
            const value = parseInt(currencyDetail.match(/(\d+)/));
            if (inventory[currencyLongName]) {
                throw Error(`Currency '${currencyLongName}' was provided twice`);
            }
            inventory[currencyLongName] = value;
        })
        this.inventory = inventory;
    }

    optimize() {
        Object.keys(CONVERSION).forEach((fromCurrency) => {
            if (this.inventory[fromCurrency] === undefined) {
                return;
            }
            Object.keys(CONVERSION[fromCurrency]).forEach((toCurrency) => {
                const conversionRate = CONVERSION[fromCurrency][toCurrency];
                const quotient = Math.floor(this.inventory[fromCurrency] / conversionRate);
                const remainder = this.inventory[fromCurrency] - (quotient * conversionRate);
                if (this.inventory[toCurrency] === undefined) {
                    this.inventory[toCurrency] = 0;
                }
                this.inventory[toCurrency] += quotient;
                this.inventory[fromCurrency] = remainder;
            });
        });
    }

    getConversion() {
        return this.inventory;
    }

}

module.exports = {
    CurrencyConverter: CurrencyConverter,
};

File: currency-exchange.py
#! /usr/local/bin/python3

from enum import Enum, auto


class CurrencyType(Enum):
    COPPER = auto()
    SILVER = auto()
    ELECTRUM = auto()
    GOLD = auto()
    PLATINUM = auto()


CONVERSION = {
    CurrencyType.COPPER: {
        CurrencyType.PLATINUM: 1000,
        CurrencyType.GOLD: 100,
        CurrencyType.ELECTRUM: 50,
        CurrencyType.SILVER: 10,
    },
    CurrencyType.SILVER: {
        CurrencyType.PLATINUM: 100,
        CurrencyType.GOLD: 10,
        CurrencyType.ELECTRUM: 5,
    },
    CurrencyType.ELECTRUM: {
        CurrencyType.PLATINUM: 20,
        CurrencyType.GOLD: 2,
    },
    CurrencyType.GOLD: {
        CurrencyType.PLATINUM: 10,
    },
}


class Currency:

    inventory = {}

    def __init__(self, inventory):
        self.inventory = inventory

    def __str__(self):
        return "\n".join([("%s: %s" % (kvp.name, self.inventory[kvp])) for kvp in self.inventory])

    def reduce(self):
        for from_currency in CONVERSION:
            if from_currency not in self.inventory:
                continue
            for to_currency in CONVERSION[from_currency]:
                conversion_rate = CONVERSION[from_currency][to_currency]
                quotient = self.inventory[from_currency] // conversion_rate
                remainder = self.inventory[from_currency] - \
                    (quotient * conversion_rate)
                if to_currency not in self.inventory:
                    self.inventory[to_currency] = 0
                self.inventory[to_currency] += quotient
                self.inventory[from_currency] = remainder
                


currency = Currency({CurrencyType.COPPER: 0, CurrencyType.SILVER: 67, CurrencyType.ELECTRUM: 0, CurrencyType.GOLD: 1933, CurrencyType.PLATINUM: 903})
currency.reduce()

print(currency)
